x-pack/plugins/security/public/session/session_timeout.ts

```diff
@@ -0,0 +1,342 @@
+    this.isVisible = isVisible;
+    if (isVisible) {
+      this.handleUserActivity();
+    }
+  };
+
+  private resetTimers = ({ expiresInMs }: SessionState) => {
+    this.stopRefreshTimer = this.stopRefreshTimer?.();
+    this.stopWarningTimer = this.stopWarningTimer?.();
+    this.stopLogoutTimer = this.stopLogoutTimer?.();
+
+    if (expiresInMs !== null) {
+      const refreshTimeout = expiresInMs - GRACE_PERIOD_MS - WARNING_MS - SESSION_CHECK_MS;
+      const warningTimeout = Math.max(expiresInMs - GRACE_PERIOD_MS - WARNING_MS, 0);
+      const logoutTimeout = Math.max(expiresInMs - GRACE_PERIOD_MS, 0);
+
+      // 1. Refresh session info before displaying any warnings
+      if (refreshTimeout > 0) {
+        this.stopRefreshTimer = startTimer(this.fetchSessionInfo, refreshTimeout);
+      }
+
+      // 2. Afterwards, show warning toast
+      if (warningTimeout > 0) {
+        this.hideWarning();
+      }
+      this.stopWarningTimer = startTimer(this.showWarning, warningTimeout);
+
+      // 3. Finally, logout
+      this.stopLogoutTimer = startTimer(() => this.sessionExpired.logout(), logoutTimeout);
+    }
+  };
+
+  private toggleEventHandlers = ({ expiresInMs, canBeExtended }: SessionState) => {
+    if (expiresInMs !== null) {
+      if (!this.channel) {
+        // Subscribe to a broadcast channel for session timeout messages.
+        // This allows us to synchronize the UX across tabs and avoid repetitive API calls.
+        this.channel = new BroadcastChannel(`${this.tenant}/session_timeout`, {
+          webWorkerSupport: false,
+        });
+        this.channel.onmessage = this.handleChannelMessage;
+      }
+
+      // Monitor activity if session can be extended. No need to do it if idleTimeout hasn't been
+      // configured.
+      if (canBeExtended && !this.stopActivityMonitor) {
+        this.stopActivityMonitor = monitorActivity(this.handleUserActivity);
+      }
+
+      // Intercept HTTP requests if session can expire
+      if (!this.removeHttpInterceptor) {
+        this.removeHttpInterceptor = this.http.intercept({
+          request: this.handleHttpRequest,
+          response: this.handleHttpResponse,
+        });
+      }
+
+      if (!this.stopVisibilityMonitor) {
+        this.stopVisibilityMonitor = monitorVisibility(this.handleVisibilityChange);
+      }
+    } else {
+      if (this.channel) {
+        this.channel.close();
+        this.channel = undefined;
+      }
+      this.removeHttpInterceptor = this.removeHttpInterceptor?.();
+      this.stopActivityMonitor = this.stopActivityMonitor?.();
+      this.stopVisibilityMonitor = this.stopVisibilityMonitor?.();
+    }
+  };
+
+  private shouldExtend() {
+    const { lastExtensionTime } = this.sessionState$.getValue();
+    return (
+      !this.isExtending &&
+      !this.warningToast &&
+      Date.now() > lastExtensionTime + EXTENSION_THROTTLE_MS
+    );
+  }
+
+  private fetchSessionInfo = async (extend = false) => {
+    this.isExtending = true;
```

Comment:
Nit: this is true whether the session is actually being extended or not. Perhaps it should be renamed to `isFetchingSessionInfo`? And/or split out into a separate variable so you can track these two things independently?